src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let stateClone = { ...state };
+  const list = [];
+
+  function removeProperties(obj, keysToRemove) {
+    for (let index = 0; index < keysToRemove.length; index++) {
+      delete stateClone[keysToRemove[index]];
+      delete obj[keysToRemove[index]];
+    }
+
+    return obj;
+  }
 
+  for (let index = 0; index < actions.length; index++) {
+    const item = actions[index];
+
+    switch (item.type) {
+      case 'addProperties':
+        stateClone = { ...stateClone, ...item.extraData };
+        list.push({ ...stateClone, ...item.extraData });
+        break;
+      case 'removeProperties':
+        list.push(removeProperties({ ...stateClone }, item.keysToRemove));
+        break;
+      case 'clear':
+        stateClone = {};
+        list.push({});
+        break;
+      default:
+        break;
```

Comment:
Consider moving the 'list.push' operation outside of the switch statement to follow the 'Don't Repeat Yourself' principle, as it is repeated in every case. You can create a new state clone within each case and then push that clone to the list after the switch statement.